### PR TITLE
Transpose 3d

### PR DESCRIPTION
**Pull request by arai713**

### CODE REVIEW

File: client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+int main()
+{
+    const int N = 16;
+    const int C = 8;
+    const int D = 8;
+    const int H = 8;
+    const int W = 8;
+
+    std::vector<std::size_t> ncdhw = {N, C, D, H, W};
```
Comment (zjing14) [inline client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp:46]: Let us do a more meaningful example, i.e., NCDHW -> NDHWC

File: client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    std::vector<std::size_t> nchwd = {N, C, H, W, D};
+    // Tensor<ADataType> a(ncdhw);
+    // Tensor<BDataType> b(nchwd);
+
+    auto size = N * C * D * H * W;
+
+    // a.GenerateTensorValue(GeneratorTensor_3<ADataType>{0.0, 1.0});
+
+    std::array<ck::index_t, 5> ab_lengths{N, C, H, W, D};
```
Comment (zjing14) [inline client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp:50]: NDHWC

File: example/44_elementwise_permute/CMakeLists.txt [hunk 3]
```diff
@@ -1,2 +1,4 @@
 add_example_executable(example_elementwise_permute_4D_fp16 elementwise_permute_4D_fp16.cpp)
 add_example_executable(example_elementwise_permute_4D_fp16_2d elementwise_permute_4D_fp16_2d.cpp)
+add_example_executable(example_elementwise_permute_5D elementwise_permute_5D.cpp)
+add_example_executable(example_elementwise_permute_5D_3d elementwise_permute_5D_3d.cpp)
```
Comment (zjing14) [inline example/44_elementwise_permute/CMakeLists.txt]: The name is confusing. Why 5D_3d? Are you using 5D kernel to do 3d?

File: client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+{
+    const int N = 16;
+    const int C = 8;
+    const int D = 8;
+    const int H = 8;
+    const int W = 8;
+
+    std::vector<std::size_t> ncdhw = {N, C, D, H, W};
+    std::vector<std::size_t> nchwd = {N, C, H, W, D};
+    // Tensor<ADataType> a(ncdhw);
```
Comment (zjing14) [inline client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp]: Clean unused comments.

File: example/44_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 5]
```diff
@@ -99,14 +99,15 @@ int main()
...
               << std::endl;
-
     bool pass = true;
 
     if(do_verification)
     {
         b_device_buf.FromDevice(b.mData.data());
         Tensor<BDataType> host_b(nhwc);
         host_elementwise4D(host_b, a, PassThrough{});
+        // LogRangeAsType<float>(std::cout << "A  : ", a.mData, ",") << std::endl;
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_4D_fp16.cpp]: Unused comments

File: example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp [hunk 6]
```diff
@@ -48,11 +48,14 @@ int main()
     bool do_verification = true;
     bool time_kernel     = true;
 
-    const int N = 120;
-    const int C = 128;
+    // const int N = 120;
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp]: Same here

File: example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp [hunk 7]
```diff
@@ -22,10 +22,10 @@ using DeviceElementwisePermuteInstance =
                                                           PassThrough,
                                                           3, // NumDim_M
                                                           1, // NumDim_N
-                                                          8,
-                                                          8,
-                                                          ck::Sequence<8>,
-                                                          ck::Sequence<8>>;
+                                                          1,
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp]: Add comments to explain the tuning parameters, like above.

File: example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp [hunk 8]
```diff
@@ -107,11 +110,13 @@ int main()
 
     float gb_per_sec = num_btype / 1.E6 / ave_time;
 
+    // LogRangeAsType<float>(std::cout << "A  : ", a.mData, ",") << std::endl;
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_4D_fp16_2d.cpp]: Same here.

File: example/44_elementwise_permute/elementwise_permute_5D.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                        functor(B_nchwd(n, c, h, w, d), a_val);
+                    }
+}
+
+int main()
+{
+    bool do_verification = true;
+    bool time_kernel     = true;
+
+    std::vector<std::size_t> ncdhw = {16, 8, 8, 8, 8};
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_5D.cpp]: Use an example that transposes from NCDHW to NDHWC.

File: example/44_elementwise_permute/elementwise_permute_5D.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+using ADataType = F16;
+using BDataType = F16;
+
+using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+using DeviceElementwisePermuteInstance =
+    ck::tensor_operation::device::DeviceElementwiseImpl<ck::Tuple<ADataType>,
+                                                        ck::Tuple<BDataType>,
+                                                        PassThrough,
+                                                        5,
+                                                        8,
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_5D.cpp]: Add comments to explain the tuning parameters.

File: example/44_elementwise_permute/elementwise_permute_5D_3d.cpp
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_5D_3d.cpp:1]: What does it mean "5D_3d"? What is the different between this example and the example above?

File: example/44_elementwise_permute/elementwise_permute_5D_3d.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+using BDataType = F16;
+
+using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+using DeviceElementwisePermuteInstance =
+    ck::tensor_operation::device::DeviceElementwise3dImpl<ck::Tuple<ADataType>,
+                                                          ck::Tuple<BDataType>,
+                                                          PassThrough,
+                                                          2, // NumDim_m, {N, C}
+                                                          2, // NumDim_n, {H, W}
+                                                          1, // NumDim_k, {D}
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_5D_3d.cpp]: Add comments to explain the tuning parameters

File: example/44_elementwise_permute/elementwise_permute_5D_3d.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    bool do_verification = true;
+    bool time_kernel     = true;
+
+    const int N = 4;
+    const int C = 16;
+    const int H = 32;
+    const int W = 5;
+    const int D = 16;
+
+    std::vector<std::size_t> ncdhw = {N, C, D, H, W};
```
Comment (zjing14) [inline example/44_elementwise_permute/elementwise_permute_5D_3d.cpp]: Use the example from NCDHW to NDHWC.

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_3d.hpp [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+                return ThreadwiseTensorSliceTransfer_v2<
+                    DataType,
+                    DataType,
+                    decltype(in_grid_3d_desc_tuple[I]),
+                    decltype(thread_buffer_desc_mnk),
+                    Sequence<MPerThread, NPerThread, KPerThread>, // SliceLengths
+                    Sequence<0, 1, 2>,                            // DimAccessOrder
+                    0,                                            // SrcVectorDim
+                    1, // InScalarPerVectorSeq::At(I),                  // ScalarPerVector
+                    1, // SrcScalarStrideInVector
```
Comment (zjing14) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_3d.hpp]: Why the Vector length is hardcoded? We should expose it as a tuning parameter or calculate it based on M/N/KThread.

File: profiler/include/profiler/profile_transpose_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+                              bool time_kernel,
+                              int N,
+                              int C,
+                              int D,
+                              int H,
+                              int W)
+{
+    bool pass = true;
+
+    std::vector<std::size_t> ncdhw = {N, C, D, H, W};
```
Comment (zjing14) [inline profiler/include/profiler/profile_transpose_impl.hpp]: Use the example from NCDHW to NDHWC

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_3d.hpp [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+                    DataType,
+                    decltype(thread_buffer_desc_mnk),
+                    decltype(out_grid_3d_desc_tuple[I]),
+                    PassThroughOp,
+                    Sequence<MPerThread, NPerThread, KPerThread>, // SliceLengths
+                    Sequence<0, 1, 2>,                            // DimAccessOrder
+                    1,                                            // SrcVectorDim
+                    1,                                            // OutScalarPerVectorSeq::At(I),
+                    InMemoryDataOperationEnum::Set,
+                    1,
```
Comment (zjing14) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_3d.hpp:169]: The vector length is hardcoded. We should expose it as a tuning parameters.

File: example/44_elementwise_permute/CMakeLists.txt (hunk 3 above)
Comment (arai713) [reply example/44_elementwise_permute/CMakeLists.txt]: The first one is using the 1d permute kernel, the second one is using the 3d permute kernel

File: client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp (hunk 1 above)
Comment (arai713) [reply client_example/23_elementwise_transpose/elementwise_transpose_3d.cpp:46]: Did you mean NCDHW -> NDHWC?

**Review by zjing14 — APPROVED** [review]
LGTM, can be merged after CI pass

**Pull request merged**